packages/dds/tree/src/feature-libraries/persisted-config/persistedConfig.ts

```diff
@@ -0,0 +1,336 @@
+	// SharedTree architecture needs to be able to transmit information to the indexes on what format they should write in.
+	// This means SummaryElements either need to receive the write version as input (a bit yucky, it's not sharedtreecore's concern)
+	// or be able to subscribe to
+}
+
+const upgradeKey = "upgrade" as const;
+
+interface UpgradeOp {
+	type: typeof upgradeKey;
+	config: PersistedConfig;
+}
+
+function isUpgradeOp(contents: any): contents is UpgradeOp {
+	return contents.type === upgradeKey;
+}
+
+// TODO: Test and validate this against reconnect & offline flows (will probably require minor api/implementation tweaks)
+class PersistedConfigStore implements IPersistedConfigStore {
+	// Stores the config that clients have implicitly agreed upon (i.e. the acked config with the highest protocolIteration)
+	private config: PersistedConfig;
+	// When initialized with a config that has a higher protocol iteration than the loaded document,
+	// this field holds that proposed newer protocol version.
+	// However, to avoid issues with readonly clients or affecting document's last edit time, actually submitting that
+	// upgrade request is performed lazily upon first edit.
+	// Once the proposed upgrade is submitted, this field should be cleared to avoid duplicate proposed submissions (which
+	// won't cause correctness issues, but have network overhead)
+	private configAwaitingSubmission: PersistedConfig | undefined;
+
+	// TODO: this tracking is arguably not necessary. the DDS should know if it has pending ops.
+	private pendingOpCount: number;
+
+	// Protocol iteration for the most recent upgrade which required op resubmission.
+	// This is tracked so it can be stored in the summary, so that clients joining the session while an upgrade op is in the
+	// collab window know to ignore concurrent ops. Undefined signifies that no upgrade ops were submitted in the collab window.
+	private mostRecentResubmissionSeq: number | undefined;
+
+	constructor(
+		private readonly schema: PersistedConfigSchema,
+		initialConfig: PersistedConfig, // should be provided at DDS construction time, generally will be specified by the application
+		private readonly submitLocalMessage: (content: any, localOpMetadata: unknown) => void,
+		private readonly onProtocolChange: (
+			current: PersistedConfig,
+			previous: PersistedConfig,
+		) => void,
+		private readonly reSubmitPendingOps: (config: PersistedConfig) => void,
+	) {
+		this.config = initialConfig;
+		this.pendingOpCount = 0;
+	}
+
+	tryProcessOp(
+		message: ISequencedDocumentMessage,
+		local: boolean,
+		localOpMetadata: unknown,
+	): boolean {
+		this.processMinSeq(message);
+		if (local) {
+			this.pendingOpCount--;
+		}
+		const { contents } = message;
+		if (isUpgradeOp(contents)) {
+			if (contents.config.protocolIteration > this.config.protocolIteration) {
+				// TODO: validate proposed config is understood by this version of the code, fail if not.
```

Comment:
How this fails is really important. I think we might want to have multiple failure modes (mainly crash OR become read-only and give the app an opportunity to inform the user)